datafusion/physical-expr/src/expressions/in_list.rs

```diff
@@ -714,193 +912,86 @@ impl PhysicalExpr for InListExpr {
+
+                    // let values_result = dict_to_val_array::<Int32Type>(&array).unwrap();
+                    // self.evaluate_non_dict(*value_type, values_result, list_values)
                 }
-                DataType::Timestamp(unit, _) => match unit {
-                    TimeUnit::Second => {
-                        make_contains_primitive!(
-                            array,
-                            list_values,
-                            self.negated,
-                            TimestampSecond,
-                            TimestampSecondArray
-                        )
-                    }
-                    TimeUnit::Millisecond => {
-                        make_contains_primitive!(
-                            array,
-                            list_values,
-                            self.negated,
-                            TimestampMillisecond,
-                            TimestampMillisecondArray
-                        )
-                    }
-                    TimeUnit::Microsecond => {
-                        make_contains_primitive!(
-                            array,
-                            list_values,
-                            self.negated,
-                            TimestampMicrosecond,
-                            TimestampMicrosecondArray
-                        )
-                    }
-                    TimeUnit::Nanosecond => {
-                        make_contains_primitive!(
-                            array,
-                            list_values,
-                            self.negated,
-                            TimestampNanosecond,
-                            TimestampNanosecondArray
-                        )
-                    }
-                },
-                datatype => Result::Err(DataFusionError::NotImplemented(format!(
-                    "InList does not support datatype {:?}.",
-                    datatype
-                ))),
+                _ => self.evaluate_non_dict(value_data_type, array, list_values),
             }
         }
     }
 }
 
+fn dict_to_val_array<K: ArrowDictionaryKeyType>(
+    array: &Arc<dyn Array>,
+) -> Result<Arc<dyn Array>> {
+    let dict_array = array.as_any().downcast_ref::<DictionaryArray<K>>().unwrap();
+    let mut dict_vals = Vec::with_capacity(dict_array.len());
+    for i in 0..array.len() {
+        let (values_array, values_index) = get_dict_value::<K>(&array, i);
+        // Look up value from Index
+        let value = match values_index {
+            Some(values_index) => ScalarValue::try_from_array(values_array, values_index),
+            // Entry was null, so return null
+            None => values_array.data_type().try_into(),
+        }?;
+        dict_vals.push(value);
+    }
+    let vals = ScalarValue::iter_to_array(dict_vals).unwrap();
+    Ok(vals)
+}
+
+// Return a boolean array indicating whether the value is in list_values
+fn evaluate_set(
```

Comment:
Why is this a new function, I think @alamb 's suggestion was to recurse into InListExpr::evaluate (by pulling it into a free function)

Reply:
Yeah, I was suggesting that since InList was already implemented for non dictionary types, we re-use that implementation (though I think that will require some restructuring of how `evaluate` is written